### ISSUE TITLE
v9 discussion and changes

### ISSUE THREAD

**Issue by kisvegabor** [kisvegabor/lvgl#3298]
See the [ROADMAP](https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.md).

### v9 schedule
- **December 4**: Feature stop, start updating the docs and testing
- **December 18**: Release candidate version and call to test
- **January 15**: Release v9.0

A large update was merged to `master` which involves API changes as well. You can use `master` to try out the latest changes but for production we recommend using the `release/v8.3` branch. 

Below you can see the main changes.

### Built in drivers
- The goal of these changes is to have built-in drivers right in the LVGL repo
- Now SDL (supporting window resize and multiple windows), Linux Frame buffer, and a TFT_eSPI wrapper is available
- Later display controllers (e.g. ILI9341) will be added too
- A "vendor layer" will be added as well to make it simpler to use the display or touch controllers e.g. with STM's SPI, or ESP's I2C, etc. 

### General API changes
- `lv_obj_add_event_cb` is renamed to `lv_obj_add_event`
- `lv_disp_...` is renamed to `lv_display_...`
- `lv_btn_...` is renamed to `lv_button_...`
- `lv_btnmatrix_...` is renamed to `lv_buttonmatrix_...`
- `lv_img_...` is renamed to `lv_image_...`
- `zoom` is renamed to `scale`
- `angle` is renamed to `rotation`
- `scr` is renamed to `screen`
- `act` is renamed to `active`
- `del` is renamed to `delete`
- `lv_obj_clear_flag` is renamed to `lv_obj_remove_flag`
- `lv_obj_clear_state` is renamed to `lv_obj_remove_state`

### New color format management
- `LV_IMG_CF_...` was removed and was replaced by `LV_COLOR_FORMAT_...`
- The image converter is not updated yet so you need to manually update the color format of the generated images
- Alpha 1bit, 2bit and 4bit color formats were removed because they can be replaced by Indexed 1 bit, 2 bit and 4 bit (alpha 8 bit is available) 
- `LV_COLOR_DEPTH 24` is supported for RGB888 rendering

### Display API
- `lv_disp_drv_t` and `lv_disp_draw_buf_t` was removed
- To create a display and set it up:
```c
lv_display_t * disp = lv_display_create(hor_res, ver_res)
lv_display_set_flush_cb(disp, flush_cb);
lv_display_set_draw_buffers(disp, buf1, buf2, buf_size_in_bytes, mode);
```
  - Note that now the buf size is in bytes and not pixels
  - modes can be:
     - `LV_DISPLAY_RENDER_MODE_PARTIAL` This way the buffers can be smaller then the display to save RAM. At least 1/10 sceen size buffer(s) are recommended.
    - `LV_DISPLAY_RENDER_MODE_DIRECT` The buffer(s) has to be screen sized and LVGL will render into the correct location of the buffer.  This way the buffer always contain the whole image. With 2 buffers the buffers' content are kept in sync automatically. (Old v7 behaviour)
    - `LV_DISPLAY_RENDER_MODE_FULL` Just always redraw the whole screen
- Similarly to the widgets, now you can attach events to the display too, using `lv_display_add_event()`
- `monitor_cb` is removed and a `LV_EVENT_RENDER_READY` event is fired instead
- Instead of having display background color and image, `lv_layer_bottom()` is added where any color can be set or any widget can be created. 
- For more details check out the docs [here](https://docs.lvgl.io/master/porting/display.html) and [here](https://docs.lvgl.io/master/overview/display.html).

#### Color format of the display
- The target color format can be adjusted in the display by calling `lv_display_set_color_format(disp, LV_COLOR_FORMAT_...)`
- `LV_COLOR_16_SWAP` is removed and `lv_display_set_color_format(disp, LV_COLOR_FORMAT_NATIVE_REVERSED)` can be used instead
- `disp_drv.scr_transp`  was removed and `lv_display_set_color_format(disp, LV_COLOR_FORMAT_NATIVE_ALPHA)` can be used instead
- `set_px_cb` is removed, you can add a custom color format converter with `disp->draw_ctx->buffer_convert = my_converter;`. As a reference see the [built-in converter](https://github.com/lvgl/lvgl/blob/master/src/draw/sw/lv_draw_sw.c#L108)

### Indev API
- Similarly to the display `lv_indev_drv_t` was removed and an input device can be created like this:
```c
lv_indev_t * indev = lv_indev_create();
lv_indev_set_type(indev, LV_INDEV_TYPE_...);
lv_indev_set_read_cb(indev, read_cb);
```

### Others
- `lv_msg` is removed and replaced by [lv_observer](https://docs.lvgl.io/master/others/observer.html)
- `lv_chart` ticks support was removed, [lv_scale](https://docs.lvgl.io/master/widgets/scale.html) can be used instead

---

The parallel rendering update is also merged to master on 2023.06.05. `lv_conf.h` was changed a lot too so don't forget to update it from `lv_conf_template.h`. You can read more about the parallel rendering changes [here](https://github.com/lvgl/lvgl/issues/4016#issue-1599634031).

---

I'll update this post if there will new changes or if something important is missing.

**Comment by KLVN** [comment]
- 3D transformations on images (https://forum.lvgl.io/t/rotate-image-image-in-the-z-direction/3963/6)

**Comment by kisvegabor** [comment]
Perspective distortion can be added in any minor version as it doesn't require a conceptual change. It's just an other option next to scale and rotation.

**Comment by Cuixudong** [comment]
Add control animation group to reduce the amount of independent control animation code.
for example：https://www.bilibili.com/video/BV1EV411s7tK   and   https://www.bilibili.com/video/BV1644y1z7MS

**Comment by kisvegabor** [comment]
For this I checked GSAP stagger as inspiration/reference: https://greensock.com/docs/v3/Staggers
What do you think about it?

**Comment by Cuixudong** [comment]
> For this I checked GSAP stagger as inspiration/reference: https://greensock.com/docs/v3/Staggers What do you think about it?

Looks great.

**Comment by kisvegabor** [comment]
I added a reminder about staggers

**Comment by GorgonMeducer** [comment]
Placeholder: 

- Introduce 9patch-images widgets/style
- Introduce an option to replace RGB565 with RGB565-A8 when `LV_COLOR_DEPTH` to `16` in benchmark
- Introduce a pipeline in renderer to support multi-GPUs/Accelerators, such as 2D-capable-DMAs, 2D GPUs, dedicated processor cores for 2D tasks etc.

**Comment by kisvegabor** [comment]
> Introduce an option to replace RGB565 with RGB565-A8 when LV_COLOR_DEPTH to 16 in benchmark

Can't we do it now in v8.3?

I added the other two to the list.

**Comment by GorgonMeducer** [comment]
@kisvegabor sure, we can do this in 8.3.

**Comment by kctan0805** [comment]
Our platform supports HW acceleration on image format RGB565 and ARGB8888.
If I set LV_COLOR_DEPTH as 16, the LV_IMG_CF_TRUE_COLOR_ALPHA image will be ARGB8565.
If I set LV_COLOR_DEPTH as 32, both LV_IMG_CF_TRUE_COLOR and LV_IMG_CF_TRUE_COLOR_ALPHA images will be ARGB8888.
Could we have an option to set LV_IMG_CF_TRUE_COLOR_ALPHA image as ARGB8888 format even LV_COLOR_DEPTH == 16?

**Comment by kisvegabor** [comment]
@GorgonMeducer 
See https://github.com/lvgl/lvgl/issues/3379

@kctan0805
Yes, we can create "direct formats" like `LV_IMG_CF_ARGB8888` or so.

**Comment by weblqb** [comment]
Build-in thread-safe if it is practical. Thanks! ^_^

**Comment by kisvegabor** [comment]
> Build-in thread-safe if it is practical. 

I find to it too complex to manage on our end compared to how simple it is to manage on the user's end :slightly_smiling_face:

**Comment by kisvegabor** [comment]
I merged the [ROADMAP](https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.md) with the list of ideas in the first post. 

I collected the  API breaking changes in to the "Plannel for v9" section. "Planned in general" contains ideas that can be added in any minor version therefore they not critical for v9.

Please take a look at the list and tell what you think. 

My approach is to start with with the simple tasks (naming, minor refactoring) and let the few complex ones for later (masking, color formats, label features).

**Comment by Cuixudong** [comment]
what about PicoGL?

**Comment by GorgonMeducer** [comment]
@kisvegabor did you forget the 9paches-image widgets?

**Comment by kisvegabor** [comment]
@Cuixudong For for what purpose do you mean PicoGL?

@GorgonMeducer It's already added [here](https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.md#styles).

**Comment by Mair** [comment]
@kisvegabor please  keep me updated to the progress (I'll also check in periodically). I'll start updating the course as soon as its stable 😊

**Comment by kisvegabor** [comment]
Thanks, @Mair. I think it will take at least a half year (probably more) to get close to release v9.

**Comment by kisvegabor** [comment]
I'm thinking about going back to a flatter directory structure. It means e.g. moving `extra/widgets` to `src/widgets`. The goal of this change would be make the features more visible and unified. Storing widgets in two places look odd in the practice. 

Do see any problems with it?

**Comment by pete-pjb** [comment]
Hi @kisvegabor ,

Having all the widgets in one place makes good sense and gets my vote 🙂 

Cheers,

Pete

**Comment by GorgonMeducer** [comment]
@kisvegabor Once you make this happen, please could you also update the PDSC file?

https://github.com/lvgl/lvgl/blob/61048554914eee36630f492ebe4e4742b51fa2c1/env_support/cmsis-pack/LVGL.lvgl.pdsc#L444-L474

**Comment by kisvegabor** [comment]
@GorgonMeducer 
Yes, no problem. 
FYI, I"ve also started to refactor `lv_conf.h` which probably affects the CMSIS pack too.

**Comment by GorgonMeducer** [comment]
@kisvegabor OK, no problem. I use the latest lv_conf_template.h whenever I create a non-alpha version. So don't worry.

**Comment by kisvegabor** [comment]
@GorgonMeducer I update the directory structure and LVGL.lvgl.pdsc too. a16e6a83

**Comment by geert-KLA-BE** [comment]
I would like to have vertical aligment of labels. Having another object to around it do that is rather wasteful.

**Comment by kisvegabor** [comment]
What is the use case when you can't simply position the label in y relative to some other object?

**Comment by geert-KLA-BE** [comment]
> What is the use case when you can't simply position the label in y relative to some other object?

For example I want a simple list of fixed heigth label buttons. I simply want to vertical align the text. This can only be done if I wrap all of them in a seperate object with a label in them. This is rather wasteful on resources just for vertical alignment of the text.

![afbeelding](https://user-images.githubusercontent.com/94625924/185079119-f19902ee-e8ba-4249-a236-df3fcda646ef.png)

**Comment by geert-KLA-BE** [comment]
Currently it seems impossible to add custom (pointer) input events. For example double tapping on a button, adding custom gestures.
Having this it would also be useful to have multi touch support. So an application could add custom gestures for that like a pinch to zoom, two finger scrolling, 2 finger rotation. It might not make sense for low end embedded devices but on powerful embedded devices it is possible to use these.

**Comment by Cuixudong** [comment]
I think the list item can be added to the left sliding menu to display the top and delete buttons, which is easier than using composite controls to achieve.
![微信截图_20220819145229](https://user-images.githubusercontent.com/23308519/185560957-48c24609-be88-4bd5-b63b-54dd5e01b40f.png)

**Comment by h13-0** [comment]
Will lvgl V9 consider supporting methods such as `lv_tileview_set_anim_time` and `lv_tileview_set_edge_flash` in lvgl V7?
The features supported by lv_tileview in V8 are not enough.

**Comment by kisvegabor** [comment]
There is an [anim](https://docs.lvgl.io/master/overview/style-props.html#anim) style property which an/should be used on more places to provide better control over the animation. I think we could use it for the tileview too.

**Comment by Jpe230** [comment]
With the removal of byte swap for RGB565, what's the recommended approach to swap bytes for displays that requires it? Swap them at a driver level?

**Comment by xiaoxiang781216** [comment]
Can we support 24bit RGB display driver interface?

**Comment by GorgonMeducer** [comment]
@Jpe230 Yes, swap hi-low bytes in `disp_flush()`

**Comment by GorgonMeducer** [comment]
@xiaoxiang781216 how many bytes per pixel? 3 or 4?

**Comment by xiaoxiang781216** [comment]
Source may any format supported by lvgl, but the display surface is 3bytes.

**Comment by kisvegabor** [comment]
> With the removal of byte swap for RGB565, what's the recommended approach to swap bytes for displays that requires it? Swap them at a driver level

In `master` there is [disp_drv->color_format](https://github.com/lvgl/lvgl/blob/master/src/hal/lv_hal_disp.h#L107) that you can set to `LV_COLOR_FORMAT_NATIVE_REVERSE`. I've tested that it's almost as fast as `LV_COLOR_16_SWAP` was in v8. 

It's not documented yet as it might still change in the future.

> Can we support 24bit RGB display driver interface?

Yes, definitely. It'd be very important to add support to it in v9. I'm not sure about the performance, as addressing 3 bytes probably won't be that fast. Converting RGBX8888 to RGB888 at the end might be faster. I'm not sure yet.

**Comment by h13-0** [comment]
# About lv_tileview
## How to enable Elastic scroll?
I want to achieve this effect:  
![chrome_iUZK0EIhY9](https://user-images.githubusercontent.com/61249472/190869144-7038dec8-2a65-4bc0-980a-be3a6899c07d.gif)

When I use
```C
  lv_obj_t* tileView = lv_tileview_create(lv_scr_act());
  lv_obj_add_flag(tileView, LV_OBJ_FLAG_SCROLL_ELASTIC);
```
but it didn't work:
![lvgl_qt_quick_YBiAO12pbj](https://user-images.githubusercontent.com/61249472/190869116-9bff2d4a-1bfa-41b8-a98d-6d044406fca9.gif)

## How to set momentum or sliding switching speed?
When I switch pages and release the mouse, I want to control the automatic homing speed of tileview.  
For example, I hope tileview is always so fast:
![lvgl_qt_quick_NsqidUAyZP](https://user-images.githubusercontent.com/61249472/190869226-511ba44e-46e3-436c-96cc-b53c6a098660.gif)

instead of:  
![lvgl_qt_quick_ewVJLF5Lof](https://user-images.githubusercontent.com/61249472/190869245-97e4796f-63c1-4378-95d4-18c31f429480.gif)

I use:
```C
  lv_obj_set_style_anim_speed(tileView, 1000, 0);
  lv_obj_set_style_anim_time(tileView, 1, 0);
  lv_obj_clear_flag(tileView, LV_OBJ_FLAG_SCROLL_MOMENTUM);
```
but nothing happend.

## My code here
```C
  lv_obj_t* tileView = lv_tileview_create(lv_scr_act());
  lv_obj_add_flag(tileView, LV_OBJ_FLAG_SCROLL_ELASTIC);
  lv_obj_set_style_anim_speed(tileView, 1000, 0);
  lv_obj_set_style_anim_time(tileView, 1, 0);
  lv_obj_clear_flag(tileView, LV_OBJ_FLAG_SCROLL_MOMENTUM);

  lv_obj_t* page1 = lv_tileview_add_tile(tileView, 0, 0, LV_DIR_BOTTOM);
  lv_obj_t* text1 = lv_label_create(page1);
  lv_label_set_text(text1, "Page 1");
  lv_obj_set_style_text_font(text1, &lv_font_montserrat_24, 0);
  
  
  lv_obj_t* page2 = lv_tileview_add_tile(tileView, 0, 1, LV_DIR_TOP | LV_DIR_RIGHT); 
  lv_obj_t* text2 = lv_label_create(page2);
  lv_label_set_text(text2, "Page 2");
  lv_obj_set_style_text_font(text2, &lv_font_montserrat_24, 0);
```
Tested on lvgl8.2 and 8.3.  
Should i change the implementation or wait for the lvgl9?
Thank you.

**Comment by xlbao123** [comment]
Support trace for easy debugging performance?

**Comment by kisvegabor** [comment]
@h13-0 
If you don't add a direction to a tile scrolling will be disabled into that direction. So you can do this:
```c
  lv_obj_t* page1 = lv_tileview_add_tile(tileView, 0, 0, LV_DIR_TOP | LV_DIR_BOTTOM);

  lv_obj_t* page2 = lv_tileview_add_tile(tileView, 0, 1, LV_DIR_BOTTOM | LV_DIR_TOP | LV_DIR_RIGHT);
```

@xlbao123 
How do you mean it in more detail?

**Comment by xlbao123** [comment]
I'm just getting started with lvgl and I see that it seems the image cache mechanism is still relatively simple, do we have any plans to provide a consistent cache management module for image and glyph bitmap

**Comment by xlbao123** [comment]
> What I want is to be able to call a unified trace callback on critical paths and branches of ui engine execution when the trace switch is turned on, so that I can access logs or a module like systrace to analyze performance bottlenecks

**Comment by GorgonMeducer** [comment]
> I'm just getting started with lvgl and I see that it seems the image cache mechanism is still relatively simple, do we have any plans to provide a consistent cache management module for image and glyph bitmap

When doing this, please keep in mind that LVGL also serves MCU with small RAM.

**Comment by sukesh-ak** [comment]
@kisvegabor What is the approx timeline for v9 release?
I have a project in ESP-IDF 4.x and planning to move to IDF 5.x
Currently the project runs with lvgl v9 but I am sure it won't be stable.

So if release timeline is far for lvgl v9, I will move back to v8.x for now.

**Comment by kisvegabor** [comment]
Still there is a lot to do so I suggest sticking with v8 :slightly_smiling_face:

**Comment by sukesh-ak** [comment]
> Still there is a lot to do so I suggest sticking with v8 🙂

Sure thank you :)

**Comment by LuWB** [comment]
In lv_draw_img_dsc_t,replace 'zoom' by 'scaling_h' and 'scaling_v'.Sometimes,lv_img need fiil object.

**Comment by kisvegabor** [comment]
I agree and it's planned. See the ROADMAP:
https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.md#styles

**Comment by LonerDan** [comment]
It could be nice to rethink and unify the configuration layer to make it easier to keep lv_conf.h in sync with Kconfig. We could perhaps use the Kconfig to define all the config options, their dependencies and default values and then use a script to generate the lv_conf_template.h/lv_conf_internal.h in a similar fashion we now generate lv_conf_internal.h from lv_conf_template.h.

**Comment by kisvegabor** [comment]
We were thinking about it but IIRC the problem was that Kconfig doesn't support "text" value type. E.g. for
```c
#define LV_MEM_CUSTOM_MALLOC    malloc
```

So `lv_conf.h` seems superior and we could generate Kconfig from it. We "just" need to parse the decoration comments to identify the levels for Kconfig. In theory it could work but in the practice a lot of custom hacks were required to have even a hand written Kconfig file.

Anyway, maybe we just need someone who spends some time with to find a solution.  Would you be interested to look into it? :slightly_smiling_face:

**Comment by LonerDan** [comment]
> We were thinking about it but IIRC the problem was that Kconfig doesn't support "text" value type. E.g. for

Per the format definition, Kconfig supports (among others) 'string' type, so it should be usable afaik
https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html

**Comment by sukesh-ak** [comment]
> We were thinking about it but IIRC the problem was that Kconfig doesn't support "text" value type. E.g. for
> 
> ```c
> #define LV_MEM_CUSTOM_MALLOC    malloc
> ```
> 
> So `lv_conf.h` seems superior and we could generate Kconfig from it. We "just" need to parse the decoration comments to identify the levels for Kconfig. In theory it could work but in the practice a lot of custom hacks were required to have even a hand written Kconfig file.
> 
> Anyway, maybe we just need someone who spends some time with to find a solution. Would you be interested to look into it? 🙂

It can be old school `ini` file format too. The one mostly used in Linux.
I use this but its `C++` though
https://github.com/Rookfighter/inifile-cpp

**Comment by sukesh-ak** [comment]
kconfig is difficult usually to navigate when there are more settings in it.
Header file is easy to do text search.

**Comment by xiaoxiang781216** [comment]
Kconfig can describe the dependence, choice, if/endif and comment/help, it's more expressible than the simple ini file. I would suggest that the reverse path which convert Kconfig to other file format instead.

**Comment by LonerDan** [comment]
> kconfig is difficult usually to navigate when there are more settings in it. Header file is easy to do text search.

I don't see why the header file would be any easier to search through than Kconfig file. Both are well-structured text files. It also doesn't make sense to introduce any new format, since the Kconfig won't probably go away when it's used at least by esp-idf afaik.

**Comment by sukesh-ak** [comment]
> > kconfig is difficult usually to navigate when there are more settings in it. Header file is easy to do text search.
> 
> I don't see why the header file would be any easier to search through than Kconfig file. Both are well-structured text files. It also doesn't make sense to introduce any new format, since the Kconfig won't probably go away when it's used at least by esp-idf afaik.

I use kconfig for ESP-IDF projects. 
But difficulty I mentioned is about navigating through menu system. That's how people usually use.

I am not suggesting to move to `ini` file but header is easier in the current context for `lvgl`.
kconfig when its touched, runs a full compile of the project so I don't usually prefer for frequently changed config elements.
Though, in the case of `lvgl`, frequent changes happen only while testing perf tweaks.

**Comment by LonerDan** [comment]
> > > kconfig is difficult usually to navigate when there are more settings in it. Header file is easy to do text search.
> > 
> > 
> > I don't see why the header file would be any easier to search through than Kconfig file. Both are well-structured text files. It also doesn't make sense to introduce any new format, since the Kconfig won't probably go away when it's used at least by esp-idf afaik.
> 
> I use kconfig for ESP-IDF projects. But difficulty I mentioned is about navigating through menu system. That's how people usually use.
> 

The tooling around Kconfig is extensive and the most commonly used 'menuconfig' is not the only way to create/modify config. There are available multiple graphical frontends that even allow you to search and other advanced features. Have a look at kconfig-frontends project.

> I am not suggesting to move to `ini` file but header is easier in the current context for `lvgl`. kconfig when its touched, runs a full compile of the project so I don't usually prefer for frequently changed config elements. Though, in the case of `lvgl`, frequent changes happen only while testing perf tweaks.

Kconfig only creates config file with defines. What you do with it next is up to you. With a good integration with the build system you use, you can rebuild only the files affected by the config change.

**Comment by kisvegabor** [comment]
I also find `.h` simpler but as a large number of people and systems are using Kconfig we should support it as well. 

The question is from which format can we generate the other.  As Konfig is not able to describe all types  we need I think we can't start from it.

**Comment by sukesh-ak** [comment]
> Kconfig only creates config file with defines. What you do with it next is up to you. With a good integration with the build system you use, you can rebuild only the files affected by the config change.

I agree but was talking about ESP-IDF.
Any changes to configuration through `menuconfig` (which uses kconfig) updates `sdkconfig` which triggers a full compile.

**Comment by LonerDan** [comment]
@kisvegabor I see what you mean, Kconfig can't produce unquoted literals without preprocessing by some other tool.
Since those #defines are predominantly used to specify custom platform abstraction (which is in most cases not that frequently changed compared to other options), we could use Kconfig configuration to specify the name of a header file that contains those definitions.
Basically, split the lvgl configuration into 'lv_conf.h' for generic options and 'lv_platform.h' with platform abstraction/customization. Kconfig then could be used to both generate 'lv_conf.h' and specify name of the platform specific header to include. We could even host 'lv_platform.h' files for common platforms.

**Comment by kisvegabor** [comment]
Something like this could work, but I can't what's the problem with generating Kconfig from `lv_conf.h`. In a header file we have infinite flexibility and we can add any markup (comments) to indicate types, hierarchy or so and our custom converter can parse these markup.

**Comment by LonerDan** [comment]
Why introduce our custom markup, create custom parser for it and introduce extra effort to maintain it when we can leverage already existing tooling and libraries ([kconfiglib](https://github.com/ulfalizer/Kconfiglib) comes to mind, although it currently deals with [its author dissapearing](https://github.com/ulfalizer/Kconfiglib/issues/121))
Also, splitting the configuration header file may be beneficial from a separation of concerns point of view, since single config file could get quite bloated with addition of new features/backends/platforms down the line.

**Comment by kisvegabor** [comment]
> Why introduce our custom markup, create custom parser for it and introduce extra effort to maintain it when we can leverage already existing tooling and libraries ([kconfiglib](https://github.com/ulfalizer/Kconfiglib) comes to mind, although it currently deals with https://github.com/ulfalizer/Kconfiglib/issues/121)

The advantage of a header file is that it's "tooling free". Just like LVGL: not special build system, tools, black magic, just add the files and compile them as you wish/can. I'd like to keep this as default, because it's the most flexible. We can build something on top of it though.

So in the end we can't avoid having one or more `.h` config files which needs to be handwritten or generated.

Anyway, part of `lv_conf.h` can be really generated from Kconfig, but still need a generator which parsed the Kconfig file create a config file from it that we like.

Splitting `lv_conf.h` to more files could also work. It seems we will need these files:
- `lv_conf_platform.h`- memory manager, big endian, etc
- `lv_conf_driver.h` - we will merge the driver in to LVGL they can have a separate file too. E.g. enable ILI934, or Linux fbdev driver. Maybe can be merged into `lv_conf_platform.h`.
- `lv_conf_features.h`  - widgets, other modules


The last two seems Kconfig compatible.

**Comment by LonerDan** [comment]
Hi @kisvegabor
What is your stance on build system integration in lvgl? There seem to be project files for Make, CMake, SConscript, whose up-to-date state seems to be based primarily on efforts of people using/needing them for their projects. I needed to integrate lvgl into my [meson](https://mesonbuild.com) based project so I created a supporting wrapper for it. One of the nice things meson allows when adding subprojects that don't use meson themselves is the ability to [add the build files for it separately, out-of-tree](https://mesonbuild.com/Wrap-dependency-system-manual.html#wrapfile-with-meson-build-patch). It also has a kind of [dependency repository](https://mesonbuild.com/Wrapdb-projects.html) containing definitions for many project (both those using meson themselves and those with community added support).
My question is, would you be open to me submitting meson support to lvgl directly or would you prefer I submit it as an out-of-tree wrap to WrapDB instead?

**Comment by kisvegabor** [comment]
Hi @LonerDan,

As I see, meson's popularity is increasing and I'm absolutely open to add meson build system support to LVGL.  

> My question is, would you be open to me submitting meson support to lvgl directly or would you prefer I submit it as an out-of-tree wrap to WrapDB instead?

I'm not familiar with WrapDB. What would be the pros and cons of having meson support directly in LVGL compared to adding it WrapDB? Does it makes sense to do both?

**Comment by LonerDan** [comment]
> I'm not familiar with WrapDB. What would be the pros and cons of having meson support directly in LVGL compared to adding it WrapDB? Does it makes sense to do both?

Yes, in case of native meson support right in lvgl it would still make sense to add it to wrapdb. We would just submit a wrap file (similar to other types of manifest files, example [here](https://github.com/mesonbuild/wrapdb/blob/master/subprojects/tracy.wrap)). That would allow users to simply add lvgl to their project by calling `meson wrap install lvgl`. I am still trying to figure out how to best handle the configuration of lvgl via meson (ie. the least hacky way that leverages strengths of the build system). I think it would work better if we had the configuration Kconfig centric as discussed previously since meson has module named 'keyval' that can load Kconfig output quite easily. I will try few experiments and let you know.

**Comment by kisvegabor** [comment]
Sounds great! :relaxed: 

Let me know if I can help with something.

**Comment by xlbao123** [comment]
Add some macros that point to user pre-defined callback in the rendering path for tracing performance. 
for example:
```
/**
 * Macro to check for existence of a registered performance tracing callback
 * and snapshots the current time in ms if the callback is valid.
 */
#define LV_PERF_START()                         					\
	{                                           					\
                if (__perf_cb) {									\
			__perf_cb(PERF_START, NULL, NULL, NULL, 0); \
		}

/**
 * Macro to stop performance trace logging and call the registered
 * performance tracing callback.
 */
#define LV_PERF_STOP(type, op, name) 								\
		if (__perf_cb) {										\
			__perf_cb(PERF_STOP, type, op, name, 0); \
		}																\
	}
```
 
Then in __perf_cb, users can record/stream these informations through logs or [standard tracing format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview). 

After all these, we can analysis performance more easily

**Comment by kisvegabor** [comment]
Makes sense. Let's get back to it when see more from the new drawing architecture.

**Comment by LonerDan** [comment]
We could try adding first-party Rust bindings for v9. There is a [third-party crate for it](https://github.com/rafaelcaricio/lvgl-rs), but it doesn't cover all of the API and seems to be currently [looking for a new maintainer](https://github.com/rafaelcaricio/lvgl-rs/commit/455a35ad7e82083ddefa748fa2376775e4d3c885).

**Comment by GorgonMeducer** [comment]
> Add some macros that point to user pre-defined callback in the rendering path for tracing performance. for example:
> 
> ```
> /**
>  * Macro to check for existence of a registered performance tracing callback
>  * and snapshots the current time in ms if the callback is valid.
>  */
> #define LV_PERF_START()                         					\
> 	{                                           					\
>                 if (__perf_cb) {									\
> 			__perf_cb(PERF_START, NULL, NULL, NULL, 0); \
> 		}
> 
> /**
>  * Macro to stop performance trace logging and call the registered
>  * performance tracing callback.
>  */
> #define LV_PERF_STOP(type, op, name) 								\
> 		if (__perf_cb) {										\
> 			__perf_cb(PERF_STOP, type, op, name, 0); \
> 		}																\
> 	}
> ```
> 
> Then in __perf_cb, users can record/stream these informations through logs or [standard tracing format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview).
> 
> After all these, we can analysis performance more easily

It is a good proposal. Once integrated, LVGL cmsis-pack users can take advantage of the (already-in-use cmsis-pack) [perf_counter](https://github.com/GorgonMeducer/perf_counter) to measure the LVGL performance in cortex-m processors.

**Comment by JAremko** [comment]
Considering multi-threaded rendering it would be great to have a way to draw canvas elements in a separate thread.  

Use case:

We have a HUD app with dynamic indicators. They're  drawn on canvas and updated pretty much every frame based on the current sensor readings. 

 It means that the rest of interface ( not related to the canvas ) isn't strongly coupled to the canvas drawing logic.  So, in theory, this is  something that can be easily done concurrently.

**Comment by JAremko** [comment]
We're currently using screens as the proposed "scene" concept.  The biggest downside to it is that we have a lot of object duplication. For example, battery indicator and timer are present in every "scene". It isn't too bad since we extended concept of LVGL screen object to have those elements but it's still a memory waste.  So scenes would be great.  :smile: 

There is one consideration about updating scene elements.   Currently, we  mark all dynamic elements with a flag and every time something might changed we walk **current** screen object three with `lv_obj_tree_walk` and issue update events:

``` c
lv_obj_tree_walk_res_t
upd_walker
  (
    lv_obj_t * obj,

    __attribute__((unused)) void * acc
  ) {
    if (lv_obj_has_flag(obj, GUI_WIDGET_DYNAMIC_TEXT)) {
      gui_widget_dynamic_text * dt_ptr =
        (gui_widget_dynamic_text * ) obj;

      if (
        dt_ptr -> allway_refresh ||
        gui_widget_dynamic_text_stale_p(dt_ptr)) {
        lv_event_send(obj, LV_EVENT_REFRESH, NULL);
      }

      return LV_OBJ_TREE_WALK_SKIP_CHILDREN;
    } else {
      // Do not walk hidden containers;
      if (lv_obj_has_flag(obj, LV_OBJ_FLAG_HIDDEN))
        return LV_OBJ_TREE_WALK_SKIP_CHILDREN;
      else
        return LV_OBJ_TREE_WALK_NEXT;
    }
  }

void gui_update_dynamic(lv_obj_t * parent) {
  lv_obj_tree_walk(parent, upd_walker, NULL);
}
```

So heaving something like `lv_obj_tree_walk` for walking scenes would be great.  

Also we extended LVGL elements to store a pointer to their  state  ( they are stateless views ) - similar to  the var binding proposal.  So `gui_widget_dynamic_text_stale_p`  takes the new state by the pointer and the one that the object used in the previous  update and if they aren't equal it knows that the object went stale.

**Comment by JAremko** [comment]
Also we extended LVGL elements to store a pointer to their  state  ( they are stateless views ) - similar to  the var binding proposal.  So `gui_widget_dynamic_text_stale_p`  takes the new state by the pointer and the one that the object used in the previous  update and if they aren't equal it knows that the object went stale.

**Comment by sukesh-ak** [comment]
> We're currently using screens as the proposed "scene" concept. The biggest downside to it is that we have a lot of object duplication. For example, battery indicator and timer are present in every "scene". It isn't too bad since we extended concept of LVGL screen object to have those elements but it's still a memory waste. So scenes would be great. 😄
> 
> There is one consideration about updating scene elements. Currently, we mark all dynamic elements with a flag and every time something might changed we walk **current** screen object three with `lv_obj_tree_walk` and issue update events:
> 
> ```c
> lv_obj_tree_walk_res_t
> upd_walker
>   (
>     lv_obj_t * obj,
> 
>     __attribute__((unused)) void * acc
>   ) {
>     if (lv_obj_has_flag(obj, GUI_WIDGET_DYNAMIC_TEXT)) {
>       gui_widget_dynamic_text * dt_ptr =
>         (gui_widget_dynamic_text * ) obj;
> 
>       if (
>         dt_ptr -> allway_refresh ||
>         gui_widget_dynamic_text_stale_p(dt_ptr)) {
>         lv_event_send(obj, LV_EVENT_REFRESH, NULL);
>       }
> 
>       return LV_OBJ_TREE_WALK_SKIP_CHILDREN;
>     } else {
>       // Do not walk hidden containers;
>       if (lv_obj_has_flag(obj, LV_OBJ_FLAG_HIDDEN))
>         return LV_OBJ_TREE_WALK_SKIP_CHILDREN;
>       else
>         return LV_OBJ_TREE_WALK_NEXT;
>     }
>   }
> 
> void gui_update_dynamic(lv_obj_t * parent) {
>   lv_obj_tree_walk(parent, upd_walker, NULL);
> }
> ```
> 
> So heaving something like `lv_obj_tree_walk` for walking scenes would be great.
> 
> Also we extended LVGL elements to store a pointer to their state ( they are stateless views ) - similar to the var binding proposal. So `gui_widget_dynamic_text_stale_p` takes the new state by the pointer and the one that the object used in the previous update and if they aren't equal it knows that the object went stale.

I had partitioned screen into 3 for this. Header, Footer and content.
And updates only content area for navigation and header has all the notification icons.

https://github.com/sukesh-ak/ESP32-TUX/blob/37822a7b6dfcb67e278c78b2aa468817b6fa0bbe/main/gui.hpp#L728

**Comment by sukesh-ak** [comment]
> Considering multi-threaded rendering it would be great to have a way to draw canvas elements in a separate thread.
> 
> Use case:
> 
> We have a HUD app with dynamic indicators. They're drawn on canvas and updated pretty much every frame based on the current sensor readings.
> 
> It means that the rest of interface ( not related to the canvas ) isn't strongly coupled to the canvas drawing logic. So, in theory, this is something that can be easily done concurrently.

Use lv_msg for this?
https://docs.lvgl.io/master/others/msg.html

**Comment by JAremko** [comment]
> I had partitioned screen into 3 for this. Header, Footer and content.
And updates only content area for navigation and header has all the notification icons.

@sukesh-ak  Makes sense.  We aren't memory constrained so I went with the whole screen switching instead. It's a simple solution since you just need to swap the root of the object tree. I also considered re-parenting top and bottom bars every time screen switches ( just move them over ) but  we ended up with the spare memory anyway.

**Comment by JAremko** [comment]
> Use lv_msg for this?
https://docs.lvgl.io/master/others/msg.html


@sukesh-ak 

You mean instead of `lv_obj_tree_walk` ? Yeah sounds like it would work.  :thinking:    We aren't on 8.3 yet and it "potentially" would be a bit more expensive. Since with `lv_obj_tree_walk` we only notify elements on the active screen and ones that don't have hidden parent.  Not sure how `lv_msg` works in this regards.

**Comment by sukesh-ak** [comment]
> > Use lv_msg for this?
> > https://docs.lvgl.io/master/others/msg.html
> 
> @sukesh-ak
> 
> You mean instead of `lv_obj_tree_walk` ? Yeah sound like it would work. 🤔 We aren't on 8.3 yet

You can check here. Multi-tasking + lv_msg being used (ESP32/ESP32-S3) + custom widget.
https://github.com/sukesh-ak/ESP32-TUX

**Comment by JAremko** [comment]
Well... I meant  filling canvas buffer with pixel data in parallel -  while the rest of LVGL interface is also updating. Something like double buffering  ( one buffer being filled by canvas API and the other one displayed as an LVGL object )  

I made an issue about it https://forum.lvgl.io/t/multi-threaded-canvas-drawing/10951

**Comment by JAremko** [comment]
For the reference [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) - same idea.

**Comment by kisvegabor** [comment]
@JAremko 

> There is one consideration about updating scene elements. Currently, we mark all dynamic elements with a flag and every time something might changed we walk current screen object three with lv_obj_tree_walk and issue update events:

I just wanted to suggest [lv_msg](https://docs.lvgl.io/master/others/msg.html) too :slightly_smiling_face: 
 
> For the reference [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas) - same idea.

We will check how it fits to our current concept.

**Comment by Cuixudong** [comment]
I learn EmberGL and run it on raspberrypi PICO,see here,https://github.com/Cuixudong/EGL_ON_PICO and https://github.com/EmberGL-org/EmberGL,but it use C++
![捕获8](https://user-images.githubusercontent.com/23308519/221783951-2b88f06c-3a8a-4d32-bd54-fb6fbedd1cdc.JPG)

**Comment by kisvegabor** [comment]
Nice! We tried to integrate [TinyGL](https://github.com/lvgl/lv_lib_tinygl) but it didn't work properly for us. 
Is EmberGL stable and easy to use?

**Comment by docterling** [comment]
so when lvgl will support 4k res?

**Comment by kisvegabor** [comment]
> so when lvgl will support 4k res?

It's already supported. The only limitation now is the max size of the images is 2047x2047 but it will be updated in v9.

**Comment by Cuixudong** [comment]
> Nice! We tried to integrate [TinyGL](https://github.com/lvgl/lv_lib_tinygl) but it didn't work properly for us. Is EmberGL stable and easy to use?

Looks easy to use,try it.

**Comment by HJQ1995A** [comment]
Do you plan to add scaling functionality to future vector graphics features to achieve a similar effect to CAD?

**Comment by kisvegabor** [comment]
@HJQ1995A I don't really understand what would be goal. Could you mention a concrete example?

**Comment by HJQ1995A** [comment]
> @HJQ1995A我真的不明白什么是目标。你能举一个具体的例子吗？

I'm sorry, my English proficiency is not very good. What I want to say is that I saw you will add SVG API later. I think SVG API should be able to draw vector graphics, but I wonder if it can support scaling and rotation functions.

**Comment by kisvegabor** [comment]
> I'm sorry, my English proficiency is not very good. 

No worries :slightly_smiling_face: 

> I wonder if it can support scaling and rotation functions.

It seems doable, but I can tell more once we get to the specification of that SVG API.

**Comment by JAremko** [comment]
With SVG drawing and scaling it could be  more straightforward  to implement something like [OpenStreetMap](https://www.openstreetmap.org/) renderer.  Map widget can be very useful for mobile devices.

**Comment by kisvegabor** [comment]
> With SVG drawing and scaling it could be more straightforward to implement something like [OpenStreetMap](https://www.openstreetmap.org/) renderer.

Does it have a generic SVG renderer too that we can reuse? Is it MCU compatible?

**Comment by JAremko** [comment]
There're plenty of [renderers](https://wiki.openstreetmap.org/wiki/Rendering) some of them can produce SVG.  Also [tools](https://wiki.openstreetmap.org/wiki/Osmconvert#Applying_Geographical_Borders) for slicing the map.

1. If you want to have a simple map of an area and render it on device (offline):
    osm (xml based format) might work fine. You can pre-convert it into svg.  For bigger maps there is a [binary format](https://wiki.openstreetmap.org/wiki/PBF_Format).

2.  Offline rendering would work to a point, eventually the map file with high [detalization for a big area](https://download.geofabrik.de/) would take too much space.

3. So probably a mix of two is the way to go. Where we receive map chunks from a server for a given location and cache them based on how much space you have.

4. I think they use int64 for ids and stuff.  Not sure if its needed or they just future proofing.

**Comment by kisvegabor** [comment]
It seems these SVG libs are mainly for MPU/mobile/desktop devices and we can't reuse them in general. Anyway, similarly to other 3rd libs/tool we can have a binding to open street map too. It's not on the roadmap now, but contribution is always welcome!

**Comment by HaoboGu** [comment]
multi-touch support?

**Comment by kisvegabor** [comment]
> multi-touch support?

For zoom, like on phones?

**Comment by Cuixudong** [comment]
> > multi-touch support?
> 
> For zoom, like on phones?

maybe,the curve also requires this function

**Comment by HaoboGu** [comment]
> For zoom, like on phones?

I would say in embedded world, dragging and swiping using multiple fingers are more useful than zoom. Imagine that you can use 2 finger swipe up&down for changing screen, and 3-finger swipe down to back to home screen, it's awesome.

I think [Macbook's Trackpad gestures](https://support.apple.com/en-us/HT204895) are good examples. How do you think that?

**Comment by kisvegabor** [comment]
> I would say in embedded world, dragging and swiping using multiple fingers are more useful than zoom. Imagine that you can use 2 finger swipe up&down for changing screen, and 3-finger swipe down to back to home screen, it's awesome.

Good idea! It might introduce a more advanced gesture system. Now we can detect only `TOP/BOTTOM/LEFT/RIGHT`.

**Comment by docterling** [comment]
> > so when lvgl will support 4k res?
> 
> It's already supported. The only limitation now is the max size of the images is 2047x2047 but it will be updated in v9.

how do i do to support it in v8.3.5

**Comment by kisvegabor** [comment]
@docterling Just set the resolution in  `lv_disp_drv_t`

**Comment by sidharth-458** [comment]
**Possible on the fly Formatting of LV_PART_ITEMS in LV_TABLE?**
Table items formatting on the fly, something similar to how **lv_label_set_recolor()** and how we use  _LV_TXT_COLOR_CMD_. Function which might change item bg color, text font color for the items inside table. Will this possible considering items are not addressable individually? 
Right now the work around is to create a container and create labels overlay on top of table, wherever we need a different format.

**Comment by kisvegabor** [comment]
@sidharth-458 
We will merge the `feat/parallel-rendering` branch soon, so I link an example from there. Here you can see how you can customize each cell: [LINK](https://github.com/lvgl/lvgl/blob/feat/parallel-rendering/examples/widgets/table/lv_example_table_1.c).

**Comment by nicusorcitu** [comment]
When is going to be the official release of LVGL 9.0? Is there any date for it?

**Comment by kisvegabor** [comment]
I still have ongoing discussion to finalize the [ROADMAP](https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.rst) for v9.0. The latest target is end of this year.

**Comment by xlbao123** [comment]
Is there a plan to acquire a markup language that enables the separation of the view presentation from the underlying code and incorporates features such as data binding based on the concept of declarative UI?

**Comment by kisvegabor** [comment]
We were talking about it [here](https://github.com/lvgl/lvgl/issues?page=3&q=is%3Aissue+is%3Aopen) (pretty long discussion) but it's quite stale now. 

Please describe why do you feel that having the markup language/declarative UI behavior  useful, especially for embedded? Do you think it's easier to work with it? Easier to maintain or understand? Or something else?

**Comment by xlbao123** [comment]
> We were talking about it [here](https://github.com/lvgl/lvgl/issues?page=3&q=is%3Aissue+is%3Aopen) (pretty long discussion) but it's quite stale now.
> 
> Please describe why do you feel that having the markup language/declarative UI behavior useful, especially for embedded? Do you think it's easier to work with it? Easier to maintain or understand? Or something else?

Yes, I have encountered similar ui engines runs on embedded mcu before, such as KoruLab (acquired by Google), Crank Software, and others. In my opinion, the benefits of markup languages in ui development are:
1. clearer code and ease of maintenance
2. easier to adapt between different resolutions
3. easy to redevelop a new set of UI based on the same underlying logic
4. reduced Firmware Size; UI can be upgraded independently
5. easier testing

Additionally, by binarizing the markup file, the impact on loading performance is minimal

**Comment by kisvegabor** [comment]
Thank you. I think the biggest problem with the declarative approach is bulding the connections the XML/JSON/etc and the real code. 

Anyway, it's not on the roadmap of v9 but we can discuss it later.

**Comment by hyansuper** [comment]
Do you think it's a good idea to make indev types user customizable? 

Maybe by adding a `proc_cb` to the indev struct, for pointer input device, it's `indev_encoder_proc`. 
No need to `lv_indev_set_type()`, just `lv_encoder_indev_create()` and a builtin encoder indev is returned.

I dont know if this will be too complecated, but the user can make indev of their own types.
I have this thought because in some rare scenarios those builtin indev types don't quite meet my needs:
1. I want to bind physical buttons to UI buttons, but UI buttons positions are not fixed (they will hide to the corner of screen when inactive), so I hope to bind directly to UI buttons instead of setting points in screen. Or sometimes there's no UI button at all.
2. I use 4 physical buttons to simulate encoder, left/right/enter/escape, I want to use escape button to get out of edit mode instead of long pressing enter button.

**Comment by kisvegabor** [comment]
It's a good idea, however we already have a lot of new features for v9.0, so I [added to the roadmap](https://github.com/lvgl/lvgl/commit/123ea43dfcef2d46da0dd7ec26fc41714e6c10bb) for the next minor release.

**Comment by jason-mao** [comment]
When will release this version?

**Comment by kisvegabor** [comment]
Our latest plan is end of this year. I've already promised it to a lot of people and companies, so I consider it final. :smile:

**Comment by calvin2021y** [comment]
will lvgl android/ios support platform native keyboard ？

**Comment by calvin2021y** [comment]
so what about desktop native keyboard ?

**Comment by kisvegabor** [comment]
Do you mean real keyboard hardware support?

**Comment by calvin2021y** [comment]
I mean the IME ( input method editor)  of desktop OS,   and OS also have the software keyboard for touch screen(or without touch screen).

**Comment by kisvegabor** [comment]
@sukesh-ak Yes, sorry for not mentioning it clearly. I've just added it to the top comment here. It's a pretty new feature so please don't hesitate to let us know if you have any improvement ideas or find any issues.

@calvin2021y 
Using the OS virtual keyboard should be possible by writing a driver for it. Similar to [this one](https://github.com/lvgl/lvgl/blob/master/src/dev/sdl/lv_sdl_keyboard.c).

**Comment by sukesh-ak** [comment]
> @sukesh-ak Yes, sorry for not mentioning it clearly. I've just added it to the top comment here. It's a pretty new feature so please don't hesitate to let us know if you have any improvement ideas or find any issues.
> 

Thanks will check it out.

**Comment by mhaberler** [comment]
played with observers - super elegant and useful!

I tried to deal with transient values which might become unavailable after a TTL (e.g. BLE beacon out of reach)

In my example I dealt with invalidation by setting a magic value (INT_MAX) in a timeout and detecting this in the observer callbackso a label or color can be set accordingly

maybe there's a more general way to convey "not available" to an observer?

what about a user_data member in lv_subject_t which observers can inspect?

**Comment by kisvegabor** [comment]
> played with observers - super elegant and useful!

Thank you for the feedback, happy that you like it :slightly_smiling_face: 
 
> it occurred to me the easiest way is to set the type to LV_SUBJECT_TYPE_NONE, notify the observers and have the client inspect the type

Interesting idea! In general it's hard to give advice about it without knowing more about the details. Anyway, it's great that you have found a pleasing solution.

**Comment by mhaberler** [comment]
it's really about conveying a value like None in Python 

sensor defective is not the same temperature as 0°  -  no value available is not the same as value zero


any specific reason why observers have a user_data field, but subjects do not?

**Comment by kisvegabor** [comment]
> sensor defective is not the same temperature as 0° - no value available is not the same as value zero

I see now. In this case I think a special value is a good idea. A more advanced solution could be to have 3 subjects:
1. for the actual value
2. for the error state
3. a subject group for both

You can subscribe to the group the get notified if there are an errors or a new value. 

> any specific reason why observers have a user_data field, but subjects do not?

Yes, oversight :smile: I've just added it in 767a44bda

**Comment by mhaberler** [comment]
yep, that would do it - need to explore subject groups

appreciate the user data - escape hatch for the weird ;)

**Comment by kisvegabor** [comment]
I've updated the [docs of the observer](https://docs.lvgl.io/master/others/observer.html#subject-groups) to describe it in more detail. It will be compiled and updated in ~10 minutes.

**Comment by djixon** [comment]
Challenge 1: Provide a simple rendering (lighting and shadowing) of all objects taking into account their Z depth, real transparency, material property (which can become part of style) on the fly. In such a model, all existing colors become diffuse colors (which means final resulting color at screen is computed based on lighting and illumination model of surface materials, that base color (or color from texture), and position of light source. Each lv_obj  can have its own light source .... just a 3D coordinate "above" (in terms of Z order with x, and y position of light source) and lighting model (Lambert, Phong, specular only, isotropic or un-isotropic). Within style property, it has material with parameters like: roughness, texture, (or simple procedural texture), index of reflection/refraction, opacity (basically alpha channel but can be changed on the fly for example if material is configured to receives shadows from "upper" layers) etc. It may looks like a huge challenging task but at simplest form it can be performed "on the fly" by actually utilizing lighting effects only over NON-FULLY-TRANSPARENT pixels (until refraction is involved). Different tricks can be used for fast calcs of any of mentioned lighting models in terms of pre-calculated tables etc. As current lv_obj are just "flat surfaces" stacked along (virtual) Z depth, for such an effect to achieve really impresive results, there is only one internal parameter required to current style elements lets call it thickness. So any border can have its own Z thickness, Padding too, PART_MAIN too, TEXT too etc. that would allow final rendering routine to properly distribute lighting effect on each part differentially based on that local Z offset of each part with respect to a Z position of object itself. Real shadowing is nothing else than taking whole rendered layer, removing chroma component from it (getting it as transparent gray-scaled image) and applying proper scaling, repositioning and blurring (depends on lighting model) with respect to position of light source, it is then applied as an overlay into shadow buffer which is propagated to "under"  layer(s) until it reach the layer which material has marked property SHADOW_RECEVER.  If material of that layer has also turned on property SHADOW_CASTER, than shadow buffer is "mixed" with shadow of that layer and propagated further until it reach layer with SHADOW_CASTER property turned off where it blends to rendered "surface" of that layer. So materials with marked SHADOW_RECEIVER property without SHADOW_CAST property will blend that shadow buffer from "above" layers into its surface buffer (which can change opacity (alpha) of that layer at those places where shadow is received). However, if SHADOW_CAST is also turned ON at such a layer, than whole shadow buffer is propagated further and actual transparency of surface of that layer is not altered by shadow.

**Comment by mhaberler** [comment]
observer: have a question on "directionality" (for the lack of a better term and before I try it out):

for example, `lv_slider_bind_value` would set the corresponding subject based on slider position 
so that is the "slider -> subject" direction

what about the other direction -"subject -> slider"?

would not make it sense to change the slider position if the  subject's value changed "externally" (eg set in application code)?

if that were the case, the subject would be the only link between UI and application
for some widgets bidirectional tracking might make sense, e.g. checkbox

disclaimer: I might be missing something fundamental - total LVGL noob here

**Comment by kisvegabor** [comment]
@mhaberler It's bidirectional and should work in both ways. If change the slider the subject will be updated. If the subject is updated the slider will be updated too.

@djixon
I also find this idea very interested but please open a new issue for it.

**Comment by mhaberler** [comment]
@kisvegabor that seriously rocks ;) 
next time I try it before posting

**Comment by kisvegabor** [comment]
No worries, I have already used the observer in a few project but it's not community approved yet. :slightly_smiling_face: 
So all feedback is very welcome. If it wasn't obvious for you it won't be for others either and it shows that the docs and comments needs to be improved.

**Comment by mhaberler** [comment]
observer and animations: 

what would be the suggested way to run an animation  on subject change?

a combination of `lv_obj_bind_state_if_eq` and `lv_objec_add_event` like given here: https://docs.lvgl.io/master/overview/anim.html#start-animation-on-an-event ?

**Comment by kisvegabor** [comment]
You can just the animation in an observer.

**Comment by kisvegabor** [comment]
I've added the schedule for the last weeks of v9  to the [roadmap](https://github.com/lvgl/lvgl/blob/master/docs/ROADMAP.rst):
### Schedule
- **December 4**: Feature stop, start updating the docs and testing
- **December 18**: Release candidate version and call to test
- **January 15**: Release v9.0

**Comment by mhaberler** [comment]
observers: compass widget adapted to use a subject group for updates

https://www.youtube.com/watch?v=K5RPykaw1eU

**Comment by kisvegabor** [comment]
> btw I found observing a subject which accidentally was not initialized produces a hang - maybe there’s a cheap test for the subject beint uninitialized and fail with a message?

We could easily check `subject->type != LV_SUBJECT_TYPE_NONE`, however earlier you have manually set the type to NONE.  What do you think?

**Comment by mhaberler** [comment]
I would add LV_SUBJECT_TYPE_INVALID as zero, and make LV_SUBJECT_TYPE_NONE > 0

**Comment by kisvegabor** [comment]
Makes sense! Would you be interested in sending a PR with this update?

**Comment by mhaberler** [comment]
yes will do

**Comment by kisvegabor** [comment]
I close this issue in favor of #4926
